core/node/node_sync/src/batch_transaction_updater/mod.rs

```diff
@@ -0,0 +1,314 @@
+    pool: ConnectionPool<Core>,
+    health_updater: HealthUpdater,
+    sleep_interval: Duration,
+}
+
+impl BatchTransactionUpdater {
+    const DEFAULT_SLEEP_INTERVAL: Duration = Duration::from_secs(5);
+
+    pub fn new(
+        sl_client: Box<dyn EthInterface>,
+        diamond_proxy_addr: Address,
+        pool: ConnectionPool<Core>,
+    ) -> Self {
+        Self::from_parts(
+            sl_client,
+            diamond_proxy_addr,
+            pool,
+            Self::DEFAULT_SLEEP_INTERVAL,
+        )
+    }
+
+    fn from_parts(
+        sl_client: Box<dyn EthInterface>,
+        diamond_proxy_addr: Address,
+        pool: ConnectionPool<Core>,
+        sleep_interval: Duration,
+    ) -> Self {
+        Self {
+            next_to_process: core::array::from_fn(|_| core::array::from_fn(|_| None)),
+            sl_client,
+            l1_transaction_verifier: L1TransactionVerifier::new(diamond_proxy_addr, pool.clone()),
+            pool,
+            health_updater: ReactiveHealthCheck::new("batch_transaction_updater").1,
+            sleep_interval,
+        }
+    }
+
+    pub fn health_check(&self) -> ReactiveHealthCheck {
+        self.health_updater.subscribe()
+    }
+
+    async fn get_block_numbers(&self) -> Result<SLBlockNumbers, EnrichedClientError> {
+        let finalized = self
+            .sl_client
+            .block(BlockId::Number(BlockNumber::Finalized))
+            .await?
+            .expect("Finalized block must be present on L1")
+            .number
+            .expect("Finalized block must contain number")
+            .as_u32()
+            .into();
+
+        let fast_finality = self
+            .sl_client
+            .block(BlockId::Number(BlockNumber::Safe))
+            .await?
+            .expect("Safe block must be present on L1")
+            .number
+            .expect("Safe block must contain number")
+            .as_u32()
+            .into();
+
+        Ok(SLBlockNumbers {
+            finalized,
+            fast_finality,
+        })
+    }
+
+    async fn apply_status_update(
+        &mut self,
+        db_eth_history_id: u32,
+        receipt: TransactionReceipt,
+        sl_block_numbers: &SLBlockNumbers,
+    ) -> anyhow::Result<()> {
+        let updated_status =
+            sl_block_numbers.get_finality_status_for_block(receipt.block_number.unwrap());
+
+        let mut connection = self
+            .pool
+            .connection_tagged("batch_transaction_updater")
+            .await?;
+
+        let db_eth_history_tx = connection
+            .eth_sender_dal()
+            .get_eth_tx_history_by_id(db_eth_history_id)
+            .await?;
+
+        let batch_number = connection
+            .blocks_dal()
+            .get_l1_batch_number_by_eth_tx_id(db_eth_history_tx.eth_tx_id)
+            .await?
+            .expect("eth_tx_history row must match a l1 batch");
+
+        // validate the transaction against db
+        let () = match db_eth_history_tx.tx_type {
```

Comment:
Why do you need `len ()=`?

Reply:
I like to do these kinds of sanity checks. It's essentially a compile time assertion that the called function returns `()`. In case it's changed, it will force revisiting the code here.

Reply:
Nobody use it like this . Ot doesn't look native